src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,152 @@ public static Path of(URI uri) {
+     * dot ('.').  If the first character in the file name string is a dot
+     * it is ignored.  If the extension cannot be determined, then an
+     * {@link Optional#empty empty} {@code Optional} is returned. This will
+     * occur if the path has zero elements ({@link #getFileName()} returns
+     * {@code null}), or the file name string does not contain a dot, or only
+     * the first character is a dot. If the last character is a dot, the
+     * extension is the empty {@code String} {@code ""}.
+     *
+     * @implSpec
+     * The default implementation is equivalent for this path to:
+     * <pre>{@code
+     *     String name = getFileName().toString();
+     *     int lastDot = name.lastIndexOf('.');
+     *     if (lastDot > 0)
+     *         lastDot == name.length() - 1) ?
+     *             Optional.of("") :
+     *             Optional.of(name.substring(lastDot + 1));
+     *     else
+     *         Optional.empty();
+     * }</pre>
+     *
+     * @return  an {@code Optional} which either contains the file name of
+     *          this path, or is {@link Optional#empty empty} if the extension
+    *           is indeterminate
+     *
+     * @since 19
+     */
+    default Optional<String> getExtension() {
+        Path fileName = getFileName();
+        if (fileName == null) {
+            return Optional.empty();
+        }
+
+        String fileNameString = fileName.toString();
+        int length = fileNameString.length();
+
+        // Indeterminate if fileNameString is too short
+        if (length > 2) {
+            int lastDotIndex = fileNameString.lastIndexOf('.');
+            // Indeterminate if no dot or only at first index
+            if (lastDotIndex > 0) {
+                return lastDotIndex == length - 1 ?
+                    Optional.of("") : // empty string if last dot at last index
+                    Optional.of(fileNameString.substring(lastDotIndex + 1));
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Returns whether the string representation of this path has an
+     * extension among the one or more extensions given as parameters.
+     * The extension of this path will be compared against the given
+     * parameters in the order in which they are supplied.
+     *
+     * @param   ext
+     *          the first extension for which to check
+     *
+     * @param   extensions
+     *          zero or more subsequent extensions for which to check
+     *
+     * @return  an {@code Optional} which either contains the found extension
+     *          or is {@link Optional#empty empty} if this path either does not
+     *          have an extension or its extension is not among those given
+     *
+     * @since 19
+     */
+    default Optional<String> hasExtension(String ext, String... extensions) {
+        String thisExtension = getExtension().orElse(null);
+        if (thisExtension == null)
+            return Optional.empty();
```

Comment:
Maybe keep around the Optional holding the extension to return if it is empty or matches; saving creating a new one.

Reply:
> Maybe keep around the Optional holding the extension to return if it is empty or matches; saving creating a new one.

I don't understand this comment. I think that this could be done in the implementation classes but I do not see how to do it here.

Reply:
Something like:
```
    default Optional<String> hasExtension(String ext, String... extensions) {
        requireNonNull(ext);
        Optional<String> extension = getExtension();
        String thisExtension = null;
        if (extension.isEmpty() || (thisExtension = extension.get()).equals(ext))
            return extension;
        ...
```

Reply:
I see. That is good. This is what I had thought of:
```
--- a/src/java.base/unix/classes/sun/nio/fs/UnixPath.java
+++ b/src/java.base/unix/classes/sun/nio/fs/UnixPath.java
+    private volatile Optional<String> theExtension = null;
+
+    @Override
+    public Optional<String> getExtension() {
+        if (theExtension == null) {
+            theExtension = Path.super.getExtension();
+        }
+        return theExtension;
+    }
```